xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Supplier;
+import io.envoyproxy.envoy.api.v2.ClusterLoadAssignment;
+import io.envoyproxy.envoy.api.v2.DiscoveryRequest;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.envoyproxy.envoy.service.discovery.v2.AggregatedDiscoveryServiceGrpc;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.stub.StreamObserver;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.CheckForNull;
+
+final class XdsClientImpl extends XdsClient {
+  static final Logger logger = Logger.getLogger(XdsClientImpl.class.getName());
+  static final String EDS_TYPE_URL =
+      "type.googleapis.com/envoy.api.v2.ClusterLoadAssignment";
+
+  private final ManagedChannel channel;
+  private final Node xdsNode;
+  private final String authority;
+  private final EndpointWatchers endpointWatchers;
+  private final AdsLifecycleManager adsLifecycleManager;
+
+  XdsClientImpl(
+      ManagedChannel channel,
+      SynchronizationContext syncCtx,
+      ScheduledExecutorService timerService,
+      BackoffPolicy.Provider backoffPolicyProvider,
+      Supplier<Stopwatch> stopwatchSupplier,
+      Node xdsNode,
+      final String authority) {
+    this.channel = checkNotNull(channel, "channel");
+    this.xdsNode = checkNotNull(xdsNode, "xdsNode");
+    this.authority = checkNotNull(authority, "authority");
+    this.endpointWatchers = new EndpointWatchers(xdsNode);
+    this.adsLifecycleManager = new AdsLifecycleManager(
+        AggregatedDiscoveryServiceGrpc.newStub(channel).withWaitForReady(),
```

Comment:
We need to send GoogleDefaultCreds (or whatever is specified in the bootstrap config) with every request, right? If we don't have all the pieces for that logic, can we have a TODO here as a reminder? I think you will pass the creds via `withCallCredentials` on the stub.

Reply:
Yes, we will do that. My PR actually initiated to do that (see https://github.com/grpc/grpc-java/pull/6320/files#diff-2ab746208e54edd7bcbba5fc408b6946R91).